astropy/utils/data.py

```diff
@@ -944,17 +969,33 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
 
     try:
         if cache:
+            clear_cache = False
+
             try:
                 # We don't need to acquire the lock here, since we are only
                 # reading
                 with _open_shelve(urlmapfn, True) as url2hash:
                     if url_key in url2hash:
                         return url2hash[url_key]
-            except ImportError:
+            except ImportError as exc:
+                msg = ('Download cache could not be read--it was saved in a '
+                       'format not supported by your Python installation '
+                       '(the {0!r} module is not installed).'.format(exc.name))
+                clear_cache = True
+            except dbm.error as exc:
+                msg = ('Download cache could not be read--{0} (this means '
+                       'your Python installation is missing the required '
+                       'db module).'.format(exc))
+                clear_cache = True
+
+            if clear_cache:
                 # Loading the cache may fail if it's in a different
                 # database format (ImportError).  In that case, delete
                 # the cache file and move on.
-                clear_download_cache()
+                cache_backups = clear_download_cache(backup=True)
+                warn(msg + '  A new cache will be created, but a backup of '
+                     'the old cache was saved to {0} and {1}.'.format(
+                     *cache_backups), AstropyWarning)
```

Comment:
In addition to making a new db file named from the db library in use, I think this can be simplified a bit to not actually make a backup of all the cached *files*.  They can just stay right where they are.  If the cleared cache DB is used then some already existing files will still be re-downloaded unnecessarily, but then once we get their hash we can just check if the file already exists and leave it there.  I think there's probably no harm in that.  The worse that can happen is that that there are saved files that are not in the DB.  We could later add a cleanup method for that if it really comes to it.